### ISSUE TITLE
How to specify setup code with Runner.bench_func() ?

### ISSUE THREAD

**Issue by parasyte** [vstinner/perf#38]
I am trying to bench some data structures with a large number of items in each structure. I would like to have a single `setup` that is called once (and only once) for all benchmark runs, because the setup takes several seconds to complete.

While debugging, I discovered this very peculiar behavior that `perf` reloads the entire main module for each run.

```python
import perf
import time

def run_benchmarks(module):
    runner = perf.Runner()

    def get_doc(attr):
        return getattr(module, attr).__doc__

    bench_names = sorted([ x for x in dir(module) if x.startswith('bench_') ], key=get_doc)
    for bench_name in bench_names:
        bench = getattr(module, bench_name)
        doc = bench.__doc__
        print('Running: %s' % (doc))
        runner.bench_func(doc, bench)

class BenchModule(object):
    def __init__(self):
        print('Creating BenchModule()')
        time.sleep(3)
        print('Created BenchModule()')

    def bench_foo(self):
        '''Benchmark: foo'''
        return list(range(1000))

    def bench_bar(self):
        '''Benchmark: bar'''
        return list(range(1000))

if __name__ == '__main__':
    print('Running benchmark')
    module = BenchModule()
    run_benchmarks(module)
```

When I run this, I see output like the following:

```
Starting benchmarks
Creating BenchModule()
Created BenchModule()
Running: Benchmark: bar
Starting benchmarks
Creating BenchModule()
Created BenchModule()
Running: Benchmark: bar
Running: Benchmark: foo
.Starting benchmarks
Creating BenchModule()
Created BenchModule()
Running: Benchmark: bar
Running: Benchmark: foo
.Starting benchmarks
Creating BenchModule()
Created BenchModule()
Running: Benchmark: bar
Running: Benchmark: foo
.Starting benchmarks
Creating BenchModule()

[ ... snip ... ]

.Starting benchmarks
Creating BenchModule()
Created BenchModule()
Running: Benchmark: bar
Running: Benchmark: foo
.
Benchmark: foo: Mean +- std dev: 9.83 us +- 0.16 us
```

... what the ? 😖

Why is it printing my `Starting benchmarks` line more than once? More importantly, *how* is it doing this? Is there some black magic going on with `subprocess(__main__)`?

But back to the original issue that lead me down this rathole; how do I prevent `perf` from running my `BenchModule` constructor on each run? I put a 3-second sleep in there to illustrate why the current behavior is obnoxious. In my real world benchmark, the setup time is more than 30 seconds, and running the complete test suite lasts a few hours.

This is possibly related: https://github.com/vstinner/perf/issues/28

**Comment by vstinner** [comment]
IMHO your benchmark works as you expect, it's just that child worker processese are flooding stdout :-)

See the documentation of the perf architecture:
http://perf.readthedocs.io/en/latest/run_benchmark.html#perf-architecture

Many processes are spawned. Even if a worker process displays "Running: Benchmark: bar" and "Running: Benchmark: foo", in practice, a worker only runs a single bench_func() function. The other methods are skipped. That's done internally by perf.

You may want to skip messages in worker processes. You can do that using:

```
args = runner.parse_args()
if not args.worker:
   # master process
   print("...")
```

Maybe I should explain that better in the documentation.

**Comment by parasyte** [comment]
Nah, it isn't flooding stdout. Only about 100 messages are printed. The problem is "needlessly" running setup code for each sample. And there are at least 60 samples for each benchmark. 30 seconds * 60 samples * 28 benchmarks = way too much time to effectively benchmark code that reports a mean less than 30ms with standard deviation of 1ms.

I don't want to skip printing messages, I want to "skip" my setup code by reusing existing data structures in memory.

**Comment by pawarren** [comment]
Is there a solution for this? My setup code takes ~20 seconds, and it looks like every single spawned process runs the setup code again instead of using the objects the parent process has already loaded into memory.

**Comment by vstinner** [comment]
We can consider to add a new parameter to bench_func() to register a "setup function".

In the meanwhile, you can work around the issue by writing a new script per benchmark no? If you want to get a single JSON file, you can use --append rather than --output.

**Comment by tyteen4a03** [comment]
Any updates on this? I'm scratching my head trying to write benchmark codes in pyperf (which includes a bunch of imports)

**Comment by cBournhonesque** [comment]
I also hit this and was surprised at how complicated it is to benchmark multiple functions inside a loop.
Other benchmarking libraries for other languages provide a simple API to do this: https://bheisler.github.io/criterion.rs/book/user_guide/benchmarking_with_inputs.html#benchmarking-with-a-range-of-values

But here it seems very convoluted to do something like
```
results = []
for x in range(..):
   results.append(runner.bench_func(name, f))
```
